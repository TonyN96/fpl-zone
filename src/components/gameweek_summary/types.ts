export interface SummaryStatType {
  label: string;
  teamCode?: number;
  playerName?: string;
  value?: string | number;
}
