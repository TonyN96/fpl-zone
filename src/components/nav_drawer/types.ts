export interface MenuItemType {
  id: string;
  label: string;
  type: "nav" | "auth";
  onClick: () => void;
}
