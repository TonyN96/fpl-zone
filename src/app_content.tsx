import React from "react";
import { useQuery } from "react-query";
import { BrowserRouter as Router,Route, Routes } from "react-router-dom";
import { useMediaQuery } from "@mui/material";
import { getAllFixtures, getGameData } from "api/fpl_api_provider";
import {
  AccountPage,
  AnalysisPage,
  FixturesAndResultsPage,
  GameweekLivePage,
  LoginPage,
  MyTeamPage,
  RegisterPage,
  ResetPage
} from "pages";
import PrivateRoute from "private_route";
import { AppData } from "types";

import { Logout } from "components/authentication";
import { Notifier, Startup } from "components/layout";

export const AppDataContext = React.createContext<AppData | null>(null);

export default function AppContent(): JSX.Element {
  // Fetching game data which will be made available throughout the app via context provider
  const {
    data: gameData,
    isError: gameDataIsError,
    error: gameDataError,
    isLoading: gameDataIsLoading
  } = useQuery("game-data", getGameData);

  // Fetching fixture data which will be made available throughout the app via context provider
  const {
    data: fixtureData,
    isError: fixtureDataIsError,
    error: fixtureDataError,
    isLoading: fixtureDataIsLoading
  } = useQuery("all-fixtures", getAllFixtures);

  const isCompact = useMediaQuery("(max-width:1500px)");

  const isLoading = gameDataIsLoading || fixtureDataIsLoading;
  const isError = gameDataIsError || fixtureDataIsError;
  const appData = gameData && fixtureData ? { ...gameData, fixtures: fixtureData, isCompact } : null;

  if (isLoading) {
    // Display loading message if data is still being fetched
    return (
      <Startup>
        <Notifier />
      </Startup>
    );
  } else if (isError) {
    // Display error message if data fetch failed
    const error = gameDataError || fixtureDataError;
    const errorMessage = error instanceof Error ? `: ${error.message}` : ".";

    return (
      <Startup>
        <Notifier type='error' message={`An error has occured: ${errorMessage}`} />
      </Startup>
    );
  }

  // Otherwise, render the app
  return (
    <AppDataContext.Provider value={appData}>
      <Router>
        <Routes>
          <Route path='*' element={<LoginPage />} />
          <Route path='/' element={<LoginPage />} />
          <Route path='/login' element={<LoginPage />} />
          <Route path='/register' element={<RegisterPage />} />
          <Route path='/reset' element={<ResetPage />} />
          <Route path='/logout' element={<Logout />} />
          <Route path='/account' element={<PrivateRoute component={<AccountPage />} />} />
          <Route
            path='/gameweek-live'
            element={<PrivateRoute component={<GameweekLivePage />} />}
          />
          <Route path='/my-team' element={<PrivateRoute component={<MyTeamPage />} />} />
          <Route
            path='/fixtures-and-results'
            element={<PrivateRoute component={<FixturesAndResultsPage />} />}
          />
          <Route path='/analysis' element={<PrivateRoute component={<AnalysisPage />} />} />
        </Routes>
      </Router>
    </AppDataContext.Provider>
  );

}
