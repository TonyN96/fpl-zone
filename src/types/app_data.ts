import { Fixture, GameSettings, Gameweek, Phase, Player, PlayerStat, Position,Team } from "types";

export interface AppData {
  gameweeks: Gameweek[];
  gameSettings: GameSettings;
  phases: Phase[];
  teams: Team[];
  playerCount: number;
  players: Player[];
  playerStats: PlayerStat[];
  positions: Position[];
  fixtures: Fixture[];
  isCompact: boolean;
}
