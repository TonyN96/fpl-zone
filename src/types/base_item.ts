import { Player } from "./player";
import { Team } from "./team";

export type BaseItem = Player | Team;

export const isPlayer = (item: BaseItem): item is Player => {
  return "web_name" in item;
};
