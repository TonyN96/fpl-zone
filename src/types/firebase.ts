export interface FirebaseResponse {
  code: string;
  message: string;
  name: string;
}
