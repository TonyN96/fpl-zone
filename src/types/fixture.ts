export interface Fixture {
  id: number;
  code: number;
  team_h: number;
  team_h_score?: unknown;
  team_a: number;
  team_a_score?: unknown;
  event: number;
  finished: boolean;
  minutes: number;
  provisional_start_time: boolean;
  kickoff_time: Date;
  event_name: string;
  is_home: boolean;
  difficulty: number;
}
