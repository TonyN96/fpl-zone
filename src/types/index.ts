export * from "./app_data";
export * from "./fixture";
export * from "./game_data";
export * from "./game_settings";
export * from "./gameweek";
export * from "./league";
export * from "./lineup";
export * from "./phase";
export * from "./player";
export * from "./player_past";
export * from "./player_performance";
export * from "./player_stat";
export * from "./player_summary";
export * from "./position";
export * from "./region";
export * from "./team";
export * from "./team_data";
export * from "./team_history";
export * from "./team_picks";
export * from "./transfer";
