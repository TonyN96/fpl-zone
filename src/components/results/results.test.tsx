import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Results from "components/results/results";
import { mockAppData } from "test";
import { Fixture, Gameweek } from "types";
import { AppDataContext } from "app_content";
import "@testing-library/jest-dom/extend-expect";

describe("Results Tests", () => {
  const latestGameweek = mockAppData.gameData.events.find(
    (gameweek) => gameweek.is_current
  ) as Gameweek;

  function createComponent(): JSX.Element {
    return (
      <AppDataContext.Provider value={mockAppData}>
        <Results />
      </AppDataContext.Provider>
    );
  }

  function clickPrevGameweekBtn(): void {
    const prevGameweekBtn = screen.getByTestId("prev-gameweek-btn");
    fireEvent.click(prevGameweekBtn);
  }

  function clickNextGameweekBtn(): void {
    const nextGameweekBtn = screen.getByTestId("next-gameweek-btn");
    fireEvent.click(nextGameweekBtn);
  }
  it("Snapshot test", () => {
    const { asFragment } = render(createComponent());
    expect(asFragment()).toMatchSnapshot();
  });

  it("Gameweek title displays as expected", () => {
    render(createComponent());

    const gameweekTitle = screen.getByTestId("selected-gameweek-title");
    expect(gameweekTitle).toHaveTextContent(latestGameweek.id.toString());
  });

  it("Gameweek fixtures rendered as expected", () => {
    render(createComponent());

    const gameweekResults: Fixture[] = mockAppData.fixtureData.filter(
      (fixture) => fixture.event === latestGameweek.id
    );

    gameweekResults.forEach((result) => {
      const resultDetails = screen.getByTestId("result-" + result.id);
      expect(resultDetails).toBeDefined();
    });
  });

  describe("Next and previous buttons", () => {
    it("Clicking prev gameweek button behaves as expected", () => {
      render(createComponent());

      clickPrevGameweekBtn();

      const gameweekTitle = screen.getByTestId("selected-gameweek-title");
      expect(gameweekTitle).toHaveTextContent((latestGameweek.id - 1).toString());
    });

    it("Clicking next gameweek button behaves as expected", () => {
      render(createComponent());

      clickNextGameweekBtn();

      const gameweekTitle = screen.getByTestId("selected-gameweek-title");
      expect(gameweekTitle).toHaveTextContent((latestGameweek.id + 1).toString());
    });

    it("Prev gameweek button disabled when selected gameweek is 1", () => {
      latestGameweek.id = 1;

      render(createComponent());

      const prevGameweekBtn = screen.getByTestId("prev-gameweek-btn");
      expect(prevGameweekBtn).toBeDisabled();
    });

    it("Next gameweek button disabled when selected gameweek is 38", () => {
      latestGameweek.id = 38;

      render(createComponent());

      const nextGameweekBtn = screen.getByTestId("next-gameweek-btn");
      expect(nextGameweekBtn).toBeDisabled();
    });
  });
});
