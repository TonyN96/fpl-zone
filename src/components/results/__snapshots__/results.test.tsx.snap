// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results Tests Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-7ghf53"
    id="results-container"
  >
    <div
      class="MuiBox-root css-gmuwbf"
    >
      <button
        aria-label="prev-gameweek"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="prev-gameweek-btn"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          aria-label="Previous gameweek"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        data-testid="selected-gameweek-title"
      >
        GAMEWEEK 33
      </h6>
      <button
        aria-label="prev-gameweek"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="next-gameweek-btn"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          aria-label="Next gameweek"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="ArrowForwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="fixture-list-container MuiBox-root css-1oy3muc"
    />
  </div>
</DocumentFragment>
`;
